File: java/meetup/src/main/java/Meetup.java
import java.time.DayOfWeek;
import java.time.LocalDate;

class Meetup {

    private int monthOfYear;
    private int year;

    Meetup(int monthOfYear, int year) {
        this.monthOfYear = monthOfYear;
        this.year = year;
    }

    LocalDate day(DayOfWeek dayOfWeek, MeetupSchedule schedule) {
        return switch(schedule) {
            case MeetupSchedule.TEENTH -> calcTeenth(dayOfWeek);
            case MeetupSchedule.LAST -> calcLast(dayOfWeek);
            default -> calcNthDayOfWeek(dayOfWeek, schedule);
        };
    }

    private LocalDate calcTeenth(DayOfWeek dayOfWeek) {
        LocalDate date = LocalDate.of(this.year, this.monthOfYear, 13);

        while (date.getDayOfWeek() != dayOfWeek) {
            date = date.plusDays(1);
        }

        return date;
    }

    private LocalDate calcLast(DayOfWeek dayOfWeek) {

        LocalDate date = LocalDate.of(this.year, this.monthOfYear, 1)
                            .plusMonths(1).minusDays(1);

        while (date.getDayOfWeek() != dayOfWeek) {
            date = date.minusDays(1);
        }

        return date;
    }

    private LocalDate calcNthDayOfWeek(DayOfWeek dayOfWeek, MeetupSchedule schedule) {

        LocalDate date = LocalDate.of(this.year, this.monthOfYear, 1);

        while (date.getDayOfWeek() != dayOfWeek) {
            date = date.plusDays(1);
        }

        int daysToAdd = switch(schedule) {
            case MeetupSchedule.FIRST -> 0;
            case MeetupSchedule.SECOND -> 7;
            case MeetupSchedule.THIRD -> 14;
            case MeetupSchedule.FOURTH -> 21;
            default -> 0;
        };

        return date.plusDays(daysToAdd);
    }
}

File: elixir/meetup/lib/meetup.ex
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """
  @dow %{monday: 1, tuesday: 2, wednesday: 3, thursday: 4, friday: 5, saturday: 6, sunday: 7}

  @type weekday ::
          :monday
          | :tuesday
          | :wednesday
          | :thursday
          | :friday
          | :saturday
          | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: Date.t()
  def meetup(year, month, weekday, schedule)
  def meetup(year, month, weekday, :teenth), do: teenth_date(year, month, weekday)
  def meetup(year, month, weekday, :first), do: date_by_descriptor(year, month, weekday, 1)
  def meetup(year, month, weekday, :second), do: date_by_descriptor(year, month, weekday, 2)
  def meetup(year, month, weekday, :third), do: date_by_descriptor(year, month, weekday, 3)
  def meetup(year, month, weekday, :fourth), do: date_by_descriptor(year, month, weekday, 4)

  def meetup(year, month, weekday, :last) do
    first_day_of_month = Date.new!(year, month, 1)

    Date.range(Date.end_of_month(first_day_of_month), first_day_of_month)
    |> Enum.filter(&(Date.day_of_week(&1) == @dow[weekday]))
    |> Enum.at(0)
  end

  defp teenth_date(year, month, weekday) do
    Date.range(Date.new!(year, month, 13), Date.new!(year, month, 19))
    |> Enum.filter(&(Date.day_of_week(&1) == @dow[weekday]))
    |> hd()
  end

  defp date_by_descriptor(year, month, weekday, descriptor) do
    first_day_of_month = Date.new!(year, month, 1)

    Date.range(first_day_of_month, Date.end_of_month(first_day_of_month))
    |> Enum.filter(&(Date.day_of_week(&1) == @dow[weekday]))
    |> Enum.at(descriptor - 1)
  end
end
